adapters/tlsio_mbedtls.c

```diff
@@ -427,76 +466,66 @@ CONCRETE_IO_HANDLE tlsio_mbedtls_create(void* io_create_parameters)
-                result->on_io_close_complete_context = NULL;
-
-                result->on_io_error = NULL;
-                result->on_io_error_context = NULL;
 
-                result->trusted_certificates = NULL;
-
-                result->socket_io = xio_create(underlying_io_interface, io_interface_parameters);
-                if (result->socket_io == NULL)
+                result->hostname = strdup(tls_io_config->hostname);
+                if (result->hostname == NULL)
                 {
-                    LogError("socket xio create failed");
                     free(result);
                     result = NULL;
                 }
                 else
                 {
-                    result->socket_io_read_bytes = NULL;
-                    result->socket_io_read_byte_count = 0;
-                    result->on_send_complete = NULL;
-                    result->on_send_complete_callback_context = NULL;
-
-                    // mbeTLS initialize
-                    mbedtls_init((void *)result, tls_io_config->hostname);
-                    result->tlsio_state = TLSIO_STATE_NOT_OPEN;
+                    result->socket_io = xio_create(underlying_io_interface, io_interface_parameters);
+                    if (result->socket_io == NULL)
+                    {
+                        LogError("socket xio create failed");
+                        free(result->hostname);
+                        free(result);
+                        result = NULL;
+                    }
+                    else
+                    {
+                        result->tls_status = 0;
+                        mbedtls_init(result);
+
+                        result->tlsio_state = TLSIO_STATE_NOT_OPEN;
+                    }
                 }
             }
         }
-        else
-        {
-            LogError("Failure allocating tlsio.");
-        }
     }
+
     return result;
 }
 
 void tlsio_mbedtls_destroy(CONCRETE_IO_HANDLE tls_io)
 {
     if (tls_io != NULL)
     {
-        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+        TLS_IO_INSTANCE *tls_io_instance = (TLS_IO_INSTANCE *)tls_io;
 
-        // mbedTLS cleanup...
-        mbedtls_ssl_free(&tls_io_instance->ssl);
-        mbedtls_ssl_config_free(&tls_io_instance->config);
-        mbedtls_x509_crt_free(&tls_io_instance->trusted_certificates_parsed);
-        mbedtls_ctr_drbg_free(&tls_io_instance->ctr_drbg);
-        mbedtls_entropy_free(&tls_io_instance->entropy);
+        mbedtls_uninit(tls_io_instance);
 
         if (tls_io_instance->socket_io_read_bytes != NULL)
         {
             free(tls_io_instance->socket_io_read_bytes);
         }
-
-        xio_destroy(tls_io_instance->socket_io);
+        if (tls_io_instance->hostname != NULL)
+        {
+            free(tls_io_instance->hostname);
+        }
         if (tls_io_instance->trusted_certificates != NULL)
         {
             free(tls_io_instance->trusted_certificates);
```

Comment:
> free(tls_io_instance->trusted_certificates); [](start = 11, length = 45)

nit: set to NULL after freeing

Reply:
fixed, but as the whole tls_io is also freed in the end, set to NULL is a little bit overkill :)